<a href="https://colab.research.google.com/github/lauravazqx/Temas-Selectos-de-Analisis-Numerico/blob/main/Proyecto_2_TSAN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto 2. Diferenciación Numérica.
# Temas Selectos de Análisis Numérico
## Elaborado por: Salvador Vázquez Laura Teresa

# Ejercicios Computacionales.

In [7]:
import numpy as np
import pandas as pd


## 11. Redactar los códigos para los siguientes métodos.

(a) Diferencias finitas hacia adelante.

(b) Diferencias finitas hacia atrás.

(c) Diferencias finitas centradas.

(d) Diferencias finitas para segunda derivada.

(e) Aproximación con polinomios de Lagrange: 3 puntos (primer forma).

(f) Aproximación con polinomios de Lagrange: 3 puntos (segunda forma).

(g) Fórmula final de 5 puntos.

(h) Fórmula final de 5 puntos (punto medio).

(i) Extrapolación de Richardson.

In [ ]:
# Respuesta para la pregunta 11.

def diferencias_finitas_adelante(f, x, h):
    return (f(x + h) - f(x)) / h

def diferencias_finitas_atras(f, x, h):
    return (f(x) - f(x - h)) / h

def diferencias_finitas_centrada(f, x, h):
    return (f(x + h) - f(x - h)) / (2 * h)

def segunda_derivada(f, x, h):
    return (f(x + h) - 2*f(x) + f(x - h)) / (h**2)

def lagrange_3_puntos_primera(f, x, h):
    return (-3*f(x) + 4*f(x + h) - f(x + 2*h)) / (2*h)

def lagrange_3_puntos_segunda(f, x, h):
    return (f(x - h) - f(x + h)) / (2*h)

def formula_5_puntos(f, x, h):
    return (-25*f(x) + 48*f(x + h) - 36*f(x + 2*h) + 16*f(x + 3*h) - 3*f(x + 4*h)) / (12*h)

def formula_5_puntos_punto_medio(f, x, h):
    return (f(x - 2*h) - 8*f(x - h) + 8*f(x + h) - f(x + 2*h)) / (12*h)

def extrapolacion_richardson(f, x, h, n=1):
    D = np.zeros((n+1, n+1))
    for i in range(n+1):
        D[i, 0] = diferencias_finitas_centrada(f, x, h / (2**i))
    for j in range(1, n+1):
        for i in range(n+1-j):
            D[i, j] = (4**j * D[i+1, j-1] - D[i, j-1]) / (4**j - 1)
    return D[0, n]

# Ejemplo de uso:
def f(x):
    return np.sin(x)

x0 = np.pi / 4
h = 0.1
print("Diferencias finitas adelante:", diferencias_finitas_adelante(f, x0, h))
print("Diferencias finitas atrás:", diferencias_finitas_atras(f, x0, h))
print("Diferencias finitas centradas:", diferencias_finitas_centrada(f, x0, h))
print("Segunda derivada:", segunda_derivada(f, x0, h))
print("Lagrange 3 puntos (primera forma):", lagrange_3_puntos_primera(f, x0, h))
print("Lagrange 3 puntos (segunda forma):", lagrange_3_puntos_segunda(f, x0, h))
print("Fórmula de 5 puntos:", formula_5_puntos(f, x0, h))
print("Fórmula de 5 puntos (punto medio):", formula_5_puntos_punto_medio(f, x0, h))
print("Extrapolación de Richardson:", extrapolacion_richardson(f, x0, h, n=2))


Diferencias finitas adelante: 0.6706029729039897
Diferencias finitas atrás: 0.7412547450958928
Diferencias finitas centradas: 0.7059288589999413
Segunda derivada: -0.706517721919031
Lagrange 3 puntos (primera forma): 0.7092790806166344
Lagrange 3 puntos (segunda forma): -0.7059288589999413
Fórmula de 5 puntos: 0.7070951994507578
Fórmula de 5 puntos (punto medio): 0.7071044269682866
Extrapolación de Richardson: 0.707106781184356


## 12. Calcular las derivadas para las siguientes funciones con cada método ($11a-11h$) y hacer una tabla comparativa por método.

(a) $f_1(x)=ln(x), x_0=1.0, h=0.4$

(b) $f_2(x)=x+e^x, x_0=0.0, h=0.7$

(c) $f_3(x)=2^xsin(x), x_0=1.05, h=0.4$

(d) $f_4(x)=x^3cos(x), x_0=2.3, h=0.5$

In [ ]:
# Respuesta para la pregunta 12.

# Definición de funciones
funciones = [
    (lambda x: np.log(x), 1.0, 0.4, 'ln(x)'),
    (lambda x: x + np.exp(x), 0.0, 0.7, 'x + e^x'),
    (lambda x: 2**x * np.sin(x), 1.05, 0.4, '2^x sin(x)'),
    (lambda x: x**3 * np.cos(x), 2.3, 0.5, 'x^3 cos(x)')
]

metodos = {
    "Diferencias Adelante": diferencias_finitas_adelante,
    "Diferencias Atrás": diferencias_finitas_atras,
    "Diferencias Centradas": diferencias_finitas_centrada,
    "Segunda Derivada": segunda_derivada,
    "Lagrange 3 ptos (1)": lagrange_3_puntos_primera,
    "Lagrange 3 ptos (2)": lagrange_3_puntos_segunda,
    "Fórmula 5 ptos": formula_5_puntos,
    "Fórmula 5 ptos (medio)": formula_5_puntos_punto_medio
}

resultados = []

for f, x0, h, nombre in funciones:
    fila = {"Función": nombre, "x0": x0, "h": h}
    for metodo, funcion in metodos.items():
        fila[metodo] = funcion(f, x0, h)
    resultados.append(fila)

# Crear DataFrame y mostrar tabla
df_resultados = pd.DataFrame(resultados)

# Estilizar tabla
def estilo(df):
    return df.style.format(precision=6).set_properties(**{'text-align': 'center'}) \
             .set_table_styles([{'selector': 'th', 'props': [('text-align', 'center')]}])

styled_df = estilo(df_resultados)
styled_df



## Tabla comparativa:

| **Método**                           | **Características y comportamiento**                                                                 |
|--------------------------------------|-----------------------------------------------------------------------------------------------------|
| **Diferencias Adelante**             | Aproximación de primer orden, error proporcional a *h*. Menor precisión; el error disminuye lentamente al reducir *h*. Se observa más error en funciones con curvatura pronunciada. |
| **Diferencias Atrás**                | Mismas características que diferencias adelante, pero usando el punto anterior. Resultados similares. |
| **Diferencias Centradas**            | Aproximación de segundo orden, error proporcional a *h²*. Mucho más precisa respecto a adelante y atrás, especialmente para funciones suaves. |
| **Segunda Derivada (centrada)**      | Cálculo de segunda derivada de orden *h²*. Resultados buenos si *h* es pequeño; el error aumenta con *h* grandes o funciones muy oscilantes. |
| **Lagrange 3 puntos (1ra derivada)** | Usa polinomio interpolante de 3 puntos. Mejor que diferencias simples, pero sensible a la elección de *h*. Precisión aceptable. |
| **Lagrange 3 puntos (2da derivada)** | Versión para segunda derivada. Resultados estables, pero menos precisos que métodos de orden mayor. |
| **Fórmula de 5 puntos**              | Aproximación de orden 4. Muy precisa; el error disminuye rápidamente. Resultados cercanos al valor exacto incluso con *h* moderado. Recomendado para alta precisión. |
| **Fórmula de 5 puntos (punto medio)**| Variante centrada, de orden alto y muy precisa. Útil cuando se busca simetría y mínima distorsión numérica. |


## 13. Utilizando Extrapolación de Richardson ($11i$) calcular las derivadas con los siguientes datos:

(a) $f_1(x)=ln(x)$ en $x_0=3$

(b) $f_2(x)=tan(x)$ en $x_0=sin^{-1}(0.8)$

(c) $f_3(x)=sin(x^2+(1/3)x)$ en $x_0=0$


In [ ]:
# Respuesta para la pregunta 13.
# Definimos la función de Extrapolación de Richardson para la primera derivada
def richardson_deriv(f, x0, h, nivel=3):
    """Extrapolación de Richardson para derivada de f en x0, con paso h y nivel especificado."""
    D = np.zeros((nivel, nivel))
    # Primera columna con aproximaciones por diferencias centradas
    for i in range(nivel):
        D[i, 0] = (f(x0 + h/(2**i)) - f(x0 - h/(2**i))) / (2 * (h/(2**i)))
    # Extrapolación de Richardson
    for j in range(1, nivel):
        for i in range(j, nivel):
            D[i, j] = D[i, j-1] + (D[i, j-1] - D[i-1, j-1]) / (4**j - 1)
    return D[nivel-1, nivel-1]

# Definición de funciones
f1 = lambda x: np.log(x)
x0_f1 = 3

f2 = lambda x: np.tan(x)
x0_f2 = np.arcsin(0.8)

f3 = lambda x: np.sin(x**2 + (1/3)*x)
x0_f3 = 0

# Paso inicial (puede ajustarse)
h = 0.1

# Cálculo de derivadas usando Extrapolación de Richardson
deriv_f1 = richardson_deriv(f1, x0_f1, h)
deriv_f2 = richardson_deriv(f2, x0_f2, h)
deriv_f3 = richardson_deriv(f3, x0_f3, h)

# Mostrar resultados
print(f"Derivada de ln(x) en x0=3: {deriv_f1}")
print(f"Derivada de tan(x) en x0=arcsin(0.8): {deriv_f2}")
print(f"Derivada de sin(x^2+(1/3)x) en x0=0: {deriv_f3}")

Derivada de ln(x) en x0=3: 0.3333333333343553
Derivada de tan(x) en x0=arcsin(0.8): 2.777778326479782
Derivada de sin(x^2+(1/3)x) en x0=0: 0.3333333333844014


## Comparación de resultados con los valores exactos:

In [10]:


# Función extrapolación de Richardson
def richardson_deriv(f, x0, h, nivel=4):
    D = np.zeros((nivel, nivel))
    for i in range(nivel):
        paso = h / (2**i)
        D[i, 0] = (f(x0 + paso) - f(x0 - paso)) / (2 * paso)
    for j in range(1, nivel):
        for i in range(j, nivel):
            D[i, j] = D[i, j-1] + (D[i, j-1] - D[i-1, j-1]) / (4**j - 1)
    return D[nivel - 1, nivel - 1]

# Funciones y valores exactos
f1 = lambda x: np.log(x)
f2 = lambda x: np.tan(x)
f3 = lambda x: np.sin(x**2 + (1/3)*x)

x0_f1 = 3
x0_f2 = np.arcsin(0.8)
x0_f3 = 0

exact_f1 = 1 / x0_f1
exact_f2 = 1 + (0.8 / np.sqrt(1 - 0.8**2))**2  # 25/9 = 2.777777...
exact_f3 = np.cos(0) * (0 + 1/3)  # 0.3333333

# Paso elegido
h = 0.1

# Aproximaciones
approx_f1 = richardson_deriv(f1, x0_f1, h)
approx_f2 = richardson_deriv(f2, x0_f2, h)
approx_f3 = richardson_deriv(f3, x0_f3, h)

# Tabla comparativa
import pandas as pd

data = {
    "Función": [r"$f_1(x)=\ln(x)$", r"$f_2(x)=\tan(x)$", r"$f_3(x)=\sin(x^2+\tfrac{x}{3})$"],
    "x0": [x0_f1, x0_f2, x0_f3],
    "Valor exacto": [exact_f1, exact_f2, exact_f3],
    "Aproximación Richardson": [approx_f1, approx_f2, approx_f3],
    "Error absoluto": [abs(exact_f1 - approx_f1), abs(exact_f2 - approx_f2), abs(exact_f3 - approx_f3)],
    "Error relativo (%)": [abs((exact_f1 - approx_f1)/exact_f1)*100, abs((exact_f2 - approx_f2)/exact_f2)*100, abs((exact_f3 - approx_f3)/exact_f3)*100]
}

df = pd.DataFrame(data)
pd.set_option('display.float_format', lambda x: f'{x:.8f}')
df


Función         x0  Valor exacto  \
0                  $f_1(x)=\ln(x)$ 3.00000000    0.33333333   
1                 $f_2(x)=\tan(x)$ 0.92729522    2.77777778   
2  $f_3(x)=\sin(x^2+\tfrac{x}{3})$ 0.00000000    0.33333333   

   Aproximación Richardson  Error absoluto  Error relativo (%)  
0               0.33333333      0.00000000          0.00000000  
1               2.77777778      0.00000000          0.00000001  
2               0.33333333      0.00000000          0.00000000

| **Función**                              | **x₀**        | **Valor exacto**     | **Aproximación (Richardson)** | **Error absoluto**    | **Error relativo (%)** |
|------------------------------------------|----------------|----------------------|-------------------------------|-----------------------|------------------------|
| \( f_1(x) = \ln(x) \)                   | 3              | 0.33333333           | (valor obtenido en el código) | (error absoluto)      | (error relativo)       |
| \( f_2(x) = \tan(x) \)                  | \( \arcsin(0.8) \approx 0.927295 \) | 2.77777778           | (valor obtenido en el código) | (error absoluto)      | (error relativo)       |
| \( f_3(x) = \sin(x^2 + \tfrac{x}{3}) \) | 0              | 0.33333333           | (valor obtenido en el código) | (error absoluto)      | (error relativo)       |


## 14. Dadas las siguientes funciones, aproximar $ \nabla f(\bar{x}_0)$ y $\nabla^2 f(\bar{x}_0)$

(a) $
f(x,y) = \cos(x+y) \sin(x^2 + y^2)$
en  $
\bar{x}_0 = (\pi, -\pi)$

(b)$
f(x,y,z) = e^{x^2 + y^2 + z^2}$
en  
$
\bar{x}_0 = (-1,2,1)$

(c)$
f(x,y,z) = \frac{xyz}{x^2 + y^2 + z^2}
$
en  $
\bar{x}_0 = (1,-1,1/2)$


Usar cualquier método visto en clase.

In [9]:
# Respuesta para la pregunta 14.
''' Usaré diferencias finitas centradas'''

def gradiente_centrada(f, x0, h=1e-5):
    n = len(x0)
    grad = np.zeros(n)
    for i in range(n):
        e = np.zeros(n)
        e[i] = 1
        grad[i] = (f(x0 + h*e) - f(x0 - h*e)) / (2*h)
    return grad

def hessiana_centrada(f, x0, h=1e-5):
    n = len(x0)
    H = np.zeros((n, n))
    for i in range(n):
        for j in range(n):
            ei = np.zeros(n)
            ej = np.zeros(n)
            ei[i] = 1
            ej[j] = 1
            if i == j:
                H[i, i] = (f(x0 + h*ei) - 2*f(x0) + f(x0 - h*ei)) / (h**2)
            else:
                H[i, j] = (f(x0 + h*ei + h*ej) - f(x0 + h*ei - h*ej) - f(x0 - h*ei + h*ej) + f(x0 - h*ei - h*ej)) / (4*h**2)
    return H

# Definimos las funciones y los puntos
funciones = [
    {
        "nombre": r"$f(x,y) = \cos(x+y) \cdot \sin(x^2+y^2)$",
        "f": lambda v: np.cos(v[0] + v[1]) * np.sin(v[0]**2 + v[1]**2),
        "x0": np.array([np.pi, -np.pi])
    },
    {
        "nombre": r"$f(x,y,z) = e^{x^2+y^2+z^2}$",
        "f": lambda v: np.exp(v[0]**2 + v[1]**2 + v[2]**2),
        "x0": np.array([-1, 2, 1])
    },
    {
        "nombre": r"$f(x,y,z) = \frac{xyz}{x^2+y^2+z^2}$",
        "f": lambda v: (v[0]*v[1]*v[2]) / (v[0]**2 + v[1]**2 + v[2]**2),
        "x0": np.array([1, -1, 0.5])
    }
]

resultados = []

for func in funciones:
    grad = gradiente_centrada(func["f"], func["x0"])
    hess = hessiana_centrada(func["f"], func["x0"])
    resultados.append({
        "Función": func["nombre"],
        "Punto": func["x0"],
        "Gradiente": grad,
        "Hessiana": hess
    })

# Convertir en DataFrame para mostrar
df = pd.DataFrame(resultados)

# Función para formatear y estilizar
def estilo(df):
    return df.style.set_properties(**{'text-align': 'center'}) \
             .format({
                 'Punto': lambda x: np.array2string(x, precision=5),
                 'Gradiente': lambda x: np.array2string(x, precision=5),
                 'Hessiana': lambda x: np.array2string(x, precision=5)
             }) \
             .set_table_styles([{'selector': 'th', 'props': [('text-align', 'center')]}])

# Mostrar DataFrame estilizado
estilo(df)
